File: 2024-php/01.php
<?php

$time_start = microtime(true);

// parse input into two sorted arrays
$input = trim(file_get_contents('01.txt'));
$a = [];
$b = [];

foreach (explode("\n", $input) as $line) {
	$parts = explode('   ', $line);
	$a[] = (int) $parts[0];
	$b[] = (int) $parts[1];
}

sort($a);
sort($b);

// add up distances between two corresponding locations
$pt1 = 0;
for ($i = 0; $i < count($a); $i++) {
	$pt1 += abs($b[$i] - $a[$i]);
}

// find number of occurences that item in $a occurs in $b
$pt2 = 0;
for ($i = 0, $j = 0; $i < count($a); $i++) {
	$occurs = 0;
	for (; $j < count($b) && $b[$j] <= $a[$i]; $j++) {
		if ($a[$i] == $b[$j]) $occurs++;
	}

	$pt2 += $a[$i] * $occurs;
}

echo "--- Day 1: Historian Hysteria ---", PHP_EOL;
echo "Part 1: ", $pt1, PHP_EOL;
echo "Part 2: ", $pt2, PHP_EOL;
echo "Took ", (microtime(true) - $time_start) * 1000, " ms", PHP_EOL;
echo PHP_EOL;

File: 2025-py/01.lua
local time_start = os.clock()
local pt1 = 0
local pt2 = 0
local p = 50

for line in io.lines() do
    local dir = string.sub(line, 0, 1)
    local amount = string.sub(line, 2)

    if dir == "L" then
        amount = -amount
    end

    p = p + amount
    pt2 = pt2 + math.abs(p // 100)
    p = p % 100

    if p == 0 then
        pt1 = pt1 + 1
    end
end

print("--- Day 1: Secret Entrance ---")
print("Part 1", pt1)
print("Part 2", pt2)
print("Took ", os.clock() - time_start, "ms")

assert(pt1 == 999, "pt1 answer is incorrect")
assert(pt2 == 6099, "pt2 answer is incorrect")
